clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -2104,32 +2151,52 @@ raw2trace_t::append_bb_entries(raw2trace_thread_data_t *tdata,
                             // dest/src of the original scatter/gather instr for all.
                             is_scatter ? instr->mem_dest_at(0) : instr->mem_src_at(0),
                             is_scatter, reg_vals, cur_pc, cur_offs, instrs_are_separate,
-                            &reached_end_of_memrefs, &interrupted))
+                            &reached_end_of_memrefs, &preempted))
                         return false;
-                    if (interrupted)
+                    if (preempted)
                         break;
                 }
             } else {
                 for (uint j = 0; j < instr->num_mem_srcs(); j++) {
                     if (!process_memref(tdata, &buf, instr, instr->mem_src_at(j), false,
                                         reg_vals, cur_pc, cur_offs, instrs_are_separate,
-                                        nullptr, &interrupted))
+                                        nullptr, &preempted))
                         return false;
-                    if (interrupted)
+                    if (preempted)
                         break;
                 }
                 // We break before subsequent memrefs on an interrupt, though with
                 // today's tracer that will never happen (i#3958).
-                for (uint j = 0; !interrupted && j < instr->num_mem_dests(); j++) {
+                for (uint j = 0; !preempted && j < instr->num_mem_dests(); j++) {
                     if (!process_memref(tdata, &buf, instr, instr->mem_dest_at(j), true,
                                         reg_vals, cur_pc, cur_offs, instrs_are_separate,
-                                        nullptr, &interrupted))
+                                        nullptr, &preempted))
                         return false;
-                    if (interrupted)
+                    if (preempted)
+                        break;
+                }
+            }
+        } else {
+            // Flush memref entries.
+            if (preempted) {
+                const offline_entry_t *next_entry = get_next_entry(tdata);
+                while (next_entry != nullptr) {
+                    if (next_entry->addr.type != OFFLINE_TYPE_MEMREF &&
+                        next_entry->addr.type != OFFLINE_TYPE_MEMREF_HIGH)
                         break;
+                    next_entry = get_next_entry(tdata);
                 }
+                if (next_entry != nullptr)
+                    tdata->pre_read.push_front(*next_entry);
             }
         }
+        // Check for a signal *after* the instruction.
+        bool interrupted = false;
+        if (!handle_kernel_interrupt_and_markers(tdata, &buf, cur_pc, cur_offs,
```

Comment:
Can we now delete the handle_kernel_interrupt_and_markers() function?  Doesn't the pre-instr "preempt" check subsume it?  If it's only needed for rseq: please rename and simplify it.